Show tooltips for `res://` and `uid://` strings in ScriptEditor

larspet opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/5027

Following the addition of tooltips (#91060), this PR adds tooltips for `res://` and `uid://` strings.

[](https://github.com/user-attachments/assets/6650c137-1a78-4fed-92ed-5ba96c0135dd)

larspet committed:
Show tooltips for res:// and uid:// strings in ScriptEditor

dalexeev commented:
1. This should work not only for `uid://` but also for `res://`.
2. For consistency with other tooltips, I suggest using the following format:
   * Format: `Resource <file name>: <Type>`
   * Example: `Resource icon.png: CompressedTexture2D`
   * Or maybe: `Resource res://icon.png: CompressedTexture2D`
   * Rationale: Documentation uses Pascal style (`var name: Type`) instead of C style (`Type name`).
   * Also, we can make the resource type links clickable. You can probably reuse `SYMBOL_HINT_ASSIGNABLE`.

larspet commented:
> 1. This should work not only for `uid://` but also for `res://`.

Makes sense, I will add that.

>     2. For consistency with other tooltips, I suggest using the following format:
>
>        * Format: `Resource <file name>: <Type>`
>        * Example: `Resource icon.png: CompressedTexture2D`
>        * Or maybe: `Resource res://icon.png: CompressedTexture2D`

`Resource <file_name>: <Type>` is probably good, but I would not want `Resource <file_path>: <Type>` because paths can get pretty long, which is why i first decided on `<Type> <file_name> <file_path>`.

I will try adding the full resource path to the tooltip body instead of having it in the title.

larspet commented:
I have now added support for `res://` as well. Since you can link to more types of files with `res://` I split everything into different types:

- `Resource`
- `Directory`
- `TextFile` (as defined by the `docks/filesystem/textfile_extensions` editor setting)
- `File` (files that are not visible in `FileSystem`, opens in File Manager instead)

dalexeev commented on editor/editor_help.cpp:
```diff
@@ -3949,6 +3981,17 @@ void EditorHelpBit::_meta_clicked(const String &p_select) {
 		} else {
 			_go_to_help(topic + ":" + symbol_class_name + ":" + link);
 		}
+	} else if (p_select.begins_with("open-ext")) {
+		String path = ProjectSettings::get_singleton()->globalize_path(p_select.substr(9));
+		OS::get_singleton()->shell_show_in_file_manager(path, true);
+	} else if (p_select.begins_with("open")) {
+		if (help_data.doc_type.type == "PackedScene") {
+			EditorNode::get_singleton()->load_scene(p_select.substr(5));
+		} else {
+			EditorNode::get_singleton()->load_resource(p_select.substr(5));
+		}
+	} else if (p_select.begins_with("show")) {
+		FileSystemDock::get_singleton()->navigate_to_path(p_select.substr(5));
```
I would suggest using a colon to more clearly denote a "pseudo-protocol" (like `open-res:`). Historically we use the symbols (`$`, `#`, `@`, `^`), but there are also real protocols `http:` and `https:`.

dalexeev commented on editor/editor_help.cpp:
```diff
@@ -3868,6 +3869,37 @@ void EditorHelpBit::_update_labels() {
+		const bool is_dir = ext.is_empty();
+		const bool is_valid = is_dir || EditorFileSystem::get_singleton()->get_valid_extensions().has(ext);
+		if (!is_dir && is_valid) {
+			content->push_meta("open " + help_data.resource_path, RichTextLabel::META_UNDERLINE_ON_HOVER);
+			content->add_image(get_editor_theme_icon(SNAME("Load")));
+			content->add_text(" " + TTR("Open"));
```
```suggestion
			content->add_image(get_editor_theme_icon(SNAME("Load")));
			content->add_text(nbsp + TTR("Open"));
			content->push_meta("open " + help_data.resource_path, RichTextLabel::META_UNDERLINE_ON_HOVER);
```

1. Use a non-breaking space.
2. Move the icon outside the link so it is not underlined.

The same below two more times.

larspet replied:
While it might look slightly strange that the icon gets underlined, I still think it's better this way since it gives you more "hit area" to click the link, and it mimics context menus more this way. It feels weird to not be able to click the icon. I did add the `nbsp` though.

larspet commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -1106,14 +1106,19 @@ void ScriptTextEditor::_show_symbol_tooltip(const String &p_symbol, int p_row, i
 		base = _find_node_for_script(base, base, script);
 	}
 
+	String doc_symbol;
 	ScriptLanguage::LookupResult result;
```
I moved the `if` to the top of the function with an early return instead, since there was nothing but a call to `show_tooltip` after the large switch. Definitely cleanest this way.

larspet commented on editor/editor_help.h:
```diff
@@ -277,6 +277,7 @@ class EditorHelpBit : public VBoxContainer {
 		String value;
 		Vector<ArgumentData> arguments;
 		String qualifiers;
+		String resource_path;
```
Originally I wanted to create the description in `parse_symbol()`, but couldn't get the links working without directly using `push_meta()` (I think I tried using `[url]` BBCode but it didn't work very well), so I had to do it in `_update_labels()` instead.

larspet commented:
I made an attempt to make the width of the tooltip match the content width, but couldn't get it working because of problems similar to #93040. I do think it would make these tooltips look much nicer (regular code-related tooltips are good as they are however), but I'll leave it for someone with more `Control` knowledge to hopefully implement in another PR.

KoBeWi commented:
I opened a very similar PR:
#101309
The difference is that it shows both path and UID for hovered files and includes the preview.
While the preview is questionable, showing full path could be useful.

KoBeWi commented:
I gave this a test and it works great, I think it's good enough for a start. There are some corner cases, like files that don't exist in FileSystem (e.g. try opening `res://.godot/editor`, the link does nothing), also I'm pretty sure users will expect this to work with relative paths and absolute paths (without `res://`), but these can be resolved later.

> Finally, I/we have forgotten about user:// paths. They would definitely benefit from tooltips as well.

Not really. `user://` does not store "permanent" files; they may exist or not, or their type might change. Showing them in tooltip does not provide any useful information. Though I suppose a fast link could be useful, so they could be supported in tooltips even for completeness sake.

KoBeWi reviewed (approved):
Needs rebase.

akien-mga commented:
I'll rebase it as this would be great to merge for 4.4.beta1 today.

Calinou commented:
Note that this would also benefit https://github.com/godotengine/godot-proposals/issues/11365. It could even use the same keyboard shortcut (since string literals can't have documentation attached to them).

akien-mga commented:
Thanks!

pennyloafers commented:
I just read the blog post, are we able to reverse search the uid to a file, as the uid has no meaning to humans and removes context to the underlying resource. I hope there is a way to quickly understand what the uid represents?

KoBeWi commented:
Well, this PR allows UID lookup from script editor. There is no other built-in tool for that currently, you either have to paste in script editor and hover or run an EditorScript that translates UID to path.

linker-err0r commented:
> I just read the blog post, are we able to reverse search the uid to a file, as the uid has no meaning to humans and removes context to the underlying resource. I hope there is a way to quickly understand what the uid represents?

This occurred to me too... It would be nice if we could alias UIDs so that we can at least render the reference itself human-readable instead of having to copy-pasta magic strings all around the project.

Tralexium commented:
Thanks for adding this great addition to the engine. I'm keen on knowing if it's currently possible to drag and drop a resource file using the UID over the absolute file path. If there's no support from that, perhaps I or someone else could make a new MR to add a new Ctrl+Shift shortcut to use the UID instead of the file path when drag & dropping a file into the code editor as it's currently really cumbersome copy paste the UID each time.

Also, I've noticed a bug where the Ctrl+Shift+Alt+C shortcut doesn't work for copying the UID when having the file selected.

KoBeWi commented:
https://github.com/godotengine/godot/pull/99094
